File: apps/haskell/src/Ch08/BankAccount.hs
module Ch08.BankAccount
    ( BankAccount
    , newBankAccount
    , getBalance
    , deposit
    , withdraw
    , transfer
    ) where

import Control.Concurrent.STM

-- | A thread-safe bank account using STM
data BankAccount = BankAccount
    { accountId      :: Int
    , accountBalance :: TVar Int
    }

-- | Create a new bank account with initial balance
newBankAccount :: Int -> Int -> IO BankAccount
newBankAccount accId initialBalance = do
    balance <- newTVarIO initialBalance
    return $ BankAccount accId balance

-- | Get the current balance
getBalance :: BankAccount -> IO Int
getBalance account = readTVarIO (accountBalance account)

-- | Deposit money into the account
deposit :: BankAccount -> Int -> IO ()
deposit account amount = atomically $
    modifyTVar' (accountBalance account) (+ amount)

-- | Withdraw money from the account
withdraw :: BankAccount -> Int -> IO Bool
withdraw account amount = atomically $ do
    balance <- readTVar (accountBalance account)
    if balance >= amount
        then do
            writeTVar (accountBalance account) (balance - amount)
            return True
        else return False

-- | Transfer money between accounts atomically (deadlock-free with STM)
transfer :: BankAccount -> BankAccount -> Int -> IO Bool
transfer from to amount = atomically $ do
    fromBalance <- readTVar (accountBalance from)
    if fromBalance >= amount
        then do
            modifyTVar' (accountBalance from) (subtract amount)
            modifyTVar' (accountBalance to) (+ amount)
            return True
        else return False

File: apps/clojure/src/grokking_concurrency/ch08/bank_account.clj
(ns grokking-concurrency.ch08.bank-account)

;; Part V - 同期と排他制御: BankAccount (STM)

(defn create-account
  "銀行口座を作成"
  [id initial-balance]
  {:id id
   :balance (ref initial-balance)})

(defn get-id
  "口座IDを取得"
  [account]
  (:id account))

(defn get-balance
  "残高を取得"
  [account]
  @(:balance account))

(defn deposit!
  "入金"
  [account amount]
  (dosync
    (alter (:balance account) + amount)))

(defn withdraw!
  "出金（残高チェック付き）"
  [account amount]
  (dosync
    (if (>= @(:balance account) amount)
      (do
        (alter (:balance account) - amount)
        true)
      false)))

(defn transfer!
  "振込（STM によるデッドロックフリー）"
  [from to amount]
  (dosync
    (if (>= @(:balance from) amount)
      (do
        (alter (:balance from) - amount)
        (alter (:balance to) + amount)
        true)
      false)))
